packages/shaders-react/src/shader-mount.tsx

```diff
@@ -128,79 +129,102 @@ export const ShaderMount: React.FC<ShaderMountProps> = forwardRef<PaperShaderEle
+    const uniformsRegistered = useRef<ShaderMountUniformsReact>(null);
+
+    // Save the initial shader fragment and context attributes, they are not allowed to change
+    const [initialShaderFragment] = useState(fragmentShaderProp);
+    const [initialContextAttrs] = useState(webGlContextAttributes);
+
+    // Initial frame that the animation starts at
+    const initialFrame = useRef(frame);
 
     const [documentVisible, setDocumentVisible] = useState(() => {
       return typeof window === 'undefined' || document.hidden === false;
     });
 
-    // Create a vanilla shader mount when we have the parent element
     useLayoutEffect(() => {
-      if (!divRef.current || shaderMount) {
-        // Dispose the shader mount if the component is unmounting
+      // Check if uniforms have changed, if yes we'll schedule processing
+      const uniformsDidChange = !fastDeepEqual(uniformsRegistered.current, uniformsProp);
+
+      if (uniformsDidChange) {
+        let cancelPromise = false;
+
+        processUniforms(uniformsProp).then((uniforms) => {
+          if (cancelPromise || !containerRef.current) {
+            return;
+          }
+
+          uniformsRegistered.current = uniformsProp;
+
+          // If shader mount already exists, we can update the uniforms
+          if (shaderMount) {
+            shaderMount.setUniforms(uniforms);
+            return;
+          }
+
+          // Otherwise, initialize a new shader mount
+          const canvas = new ShaderMountVanilla(
+            containerRef.current,
+            initialShaderFragment,
+            uniforms,
+            initialContextAttrs,
+            speed,
+            frame,
+            minPixelRatio,
+            maxPixelCount
+          );
+
+          setShaderMount(canvas);
+        });
+
         return () => {
-          shaderMount?.dispose();
+          cancelPromise = true;
         };
       }
 
-      let cancelPromise = false;
-
-      // Use initial params from a ref to keep them out of the effect deps, running on mount only
-      const { fragmentShader, uniformsProp, frame, speed, minPixelRatio, maxPixelCount, webGlContextAttributes } =
-        initialParams.current;
-
-      processUniforms(uniformsProp).then((uniforms) => {
-        if (cancelPromise || !divRef.current) {
-          return;
-        }
-
-        const canvas = new ShaderMountVanilla(
-          divRef.current,
-          fragmentShader,
-          uniforms,
-          webGlContextAttributes,
-          speed,
-          frame,
-          minPixelRatio,
-          maxPixelCount
-        );
-
-        setShaderMount(canvas);
-      });
+      // Can't do granular updates if shader mount doesn't exist yet
+      if (!shaderMount) {
+        return;
+      }
 
-      // Cancel the promise if the effect is re-run or the component is unmounting
-      return () => {
-        cancelPromise = true;
-      };
-    }, [shaderMount]);
+      if (shaderMount.speed !== speed * +documentVisible) {
```

Comment:
this is super cute 😂 I think stephen had a similar comment, but would love to see an explicit branch to turn off speed so it's explicitly obvious what's happening when document visible is false